Potential new/delete mismatch when publishing a custom type that has overloaded new/delete operators (e.g. pcl::PointCloud) using type adapters

### Generated by Generative AI

No

### Operating System:

Ubuntu 24

### ROS version or commit hash:

jazzy

### RMW implementation (if applicable):

zenoh

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

rclcpp

### 'ros2 doctor --report' output

_No response_

### Steps to reproduce issue

Using the following minimal code 
```cpp
#include <pcl/point_cloud.h>
#include <pcl/point_types.h>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>

template <class PointType>
struct rclcpp::TypeAdapter<pcl::PointCloud<PointType>, sensor_msgs::msg::PointCloud2> {
  using is_specialized = std::true_type;
  using custom_type = pcl::PointCloud<PointType>;
  using ros_message_type = sensor_msgs::msg::PointCloud2;

  static void convert_to_ros_message(const pcl::PointCloud<PointType>& source,
                                     sensor_msgs::msg::PointCloud2& destination) {}

  static void convert_to_custom(const sensor_msgs::msg::PointCloud2& source,
                                pcl::PointCloud<PointType>& destination) {}
};

int main() {
  auto node = std::make_shared<rclcpp::Node>("mismatched_new_delete_reproducer");
  auto pub = node->create_publisher<
      rclcpp::adapt_type<pcl::PointCloud<pcl::PointXYZ>>::as<sensor_msgs::msg::PointCloud2>>(
      "/point_cloud", 1);
  
  pcl::PointCloud<pcl::PointXYZ> point_cloud;
  pub->publish(point_cloud);
  
}
```
the compiler correctly detects a new/delete mismatch and reports a warning:

<details>

```

In file included from /usr/include/eigen3/Eigen/Core:166,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/ivo/colcon_ws/src/my_package/my_node.cpp:
In function ‘void Eigen::internal::aligned_free(void*)’,
    inlined from ‘void Eigen::internal::conditional_aligned_free(void*) [with bool Align = true]’ at /usr/include/eigen3/Eigen/src/Core/util/Memory.h:259:15,
    inlined from ‘static void pcl::PointCloud<PointT>::operator delete(void*) [with PointT = pcl::PointXYZ]’ at /usr/include/pcl-1.14/pcl/point_cloud.h:900:7,
    inlined from ‘void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = pcl::PointCloud<pcl::PointXYZ>]’ at /usr/include/c++/13/bits/unique_ptr.h:99:2,
    inlined from ‘std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = pcl::PointCloud<pcl::PointXYZ>; _Dp = std::default_delete<pcl::PointCloud<pcl::PointXYZ> >]’ at /usr/include/c++/13/bits/unique_ptr.h:404:17,
    inlined from ‘std::enable_if_t<(typename rclcpp::TypeAdapter<MessageT>::is_specialized::value && std::is_same<T, typename rclcpp::TypeAdapter<MessageT>::custom_type>::value)> rclcpp::Publisher<MessageT, AllocatorT>::publish(const T&) [with T = pcl::PointCloud<pcl::PointXYZ>; MessageT = rclcpp::TypeAdapter<pcl::PointCloud<pcl::PointXYZ>, sensor_msgs::msg::PointCloud2_<std::allocator<void> >, void>; AllocatorT = std::allocator<void>]’ at /opt/ros/jazzy/include/rclcpp/rclcpp/publisher.hpp:390:18,
    inlined from ‘void fsd::FSDGMMMapping::publish_sensor_model(const std::string&)’ at /home/ivo/colcon_ws/src/my_package/my_node.cpp:354:57:
/usr/include/eigen3/Eigen/src/Core/util/Memory.h:203:9: warning: ‘void free(void*)’ called on pointer returned from a mismatched allocation function [-Wmismatched-new-delete]
  203 |     free(ptr);
      |     ~~~~^~~~~
In file included from /usr/include/x86_64-linux-gnu/c++/13/bits/c++allocator.h:33,
                 from /usr/include/c++/13/bits/allocator.h:46,
                 from /usr/include/c++/13/string:43,
                 from /usr/include/c++/13/bits/locale_classes.h:40,
                 from /usr/include/c++/13/bits/ios_base.h:41,
                 from /usr/include/c++/13/ios:44,
                 from /usr/include/c++/13/istream:40,
                 from /usr/include/c++/13/sstream:40,
                 from /usr/include/c++/13/chrono:45,
                 from /usr/include/pcl-1.14/pcl/common/time.h:41,
                 from /home/ivo/colcon_ws/src/my_package/my_node.cpp:1:
In member function ‘_Tp* std::__new_allocator<_Tp>::allocate(size_type, const void*) [with _Tp = pcl::PointCloud<pcl::PointXYZ>]’,
    inlined from ‘constexpr _Tp* std::allocator< <template-parameter-1-1> >::allocate(std::size_t) [with _Tp = pcl::PointCloud<pcl::PointXYZ>]’ at /usr/include/c++/13/bits/allocator.h:198:40,
    inlined from ‘static constexpr _Tp* std::allocator_traits<std::allocator<_CharT> >::allocate(allocator_type&, size_type) [with _Tp = pcl::PointCloud<pcl::PointXYZ>]’ at /usr/include/c++/13/bits/alloc_traits.h:482:28,
    inlined from ‘std::unique_ptr<typename rclcpp::TypeAdapter<MessageT>::custom_type, typename std::conditional<std::is_same<typename std::allocator_traits<typename std::allocator_traits<_Allocator>::rebind_traits<typename rclcpp::TypeAdapter<MessageT, void, void>::custom_type>::allocator_type>::rebind_alloc<typename rclcpp::TypeAdapter<MessageT, void, void>::custom_type>, std::allocator<typename rclcpp::TypeAdapter<MessageT>::custom_type> >::value, std::default_delete<typename rclcpp::TypeAdapter<MessageT>::custom_type>, rclcpp::allocator::AllocatorDeleter<typename std::allocator_traits<_Allocator>::rebind_traits<typename rclcpp::TypeAdapter<MessageT, void, void>::custom_type>::allocator_type> >::type> rclcpp::Publisher<MessageT, AllocatorT>::duplicate_type_adapt_message_as_unique_ptr(const PublishedType&) [with MessageT = rclcpp::TypeAdapter<pcl::PointCloud<pcl::PointXYZ>, sensor_msgs::msg::PointCloud2_<std::allocator<void> >, void>; AllocatorT = std::allocator<void>]’ at /opt/ros/jazzy/include/rclcpp/rclcpp/publisher.hpp:603:54,
    inlined from ‘std::enable_if_t<(typename rclcpp::TypeAdapter<MessageT>::is_specialized::value && std::is_same<T, typename rclcpp::TypeAdapter<MessageT>::custom_type>::value)> rclcpp::Publisher<MessageT, AllocatorT>::publish(const T&) [with T = pcl::PointCloud<pcl::PointXYZ>; MessageT = rclcpp::TypeAdapter<pcl::PointCloud<pcl::PointXYZ>, sensor_msgs::msg::PointCloud2_<std::allocator<void> >, void>; AllocatorT = std::allocator<void>]’ at /opt/ros/jazzy/include/rclcpp/rclcpp/publisher.hpp:389:10,
```

</details>

Edit: This issue is only present when using the ROS-package `pcl_ros` that pulls in the dependency on PCL. This is because pcl_ros still specifies C++14 standard (causing `-D EIGEN_HAS_CXX17_OVERALIGN=0` to be present in the compiler commandline). Since C++17, the aligned allocators that PCL uses were standartized and therefore the pcl::PointCloud does not have custom new/delete operators anymore.  
This means, if you would depend directly on a recent version of PCL instead of using pcl_ros, you will not observe this problem.

### Expected behavior

rclcpp should use matching new/delete operators to prevent heap corruptions

### Actual behavior

New/delete mismatch detected at compile time in form of a warning 

### Additional information

Consider a custom type that has overloaded new and delete operators, i.e. the commonly used `pcl::PointCloud`. We can publish this type by using type adapters, i.e. as a `sensor_msgs/PointCloud2`. If we publish this type by value however, a new/delete mismatch occurs: global operator new is used together with the custom operator delete.
This occurs only if we publish by value, if we publish by `std::unique_ptr`, this issue is not present.

### Root cause 

This is due to the following `publish`-overload that first copies the given message:
https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/publisher.hpp#L357

The root cause of not using the custom operator new is that rclcpp allows for custom allocators, and how they work. If the message is published by value, it is first copied using `std::allocator_traits<A>:: allocate/construct` in the following function, which always uses the global placement new operator ([cppreference](https://en.cppreference.com/w/cpp/memory/allocator_traits/construct.html)](https://en.cppreference.com/w/cpp/memory/allocator/construct.html)):

https://github.com/ros2/rclcpp/blob/rolling/rclcpp/include/rclcpp/publisher.hpp#L575


## Proposed solution 

Add a static_assert in the `publish(const T&)` overload that asserts the type has no custom operator new. Suggest the user should use the `publish(std::unique_ptr<T>)` overload instead.

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -576,10 +576,9 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    static_assert(!detail::has_overloaded_operator_new_v<PublishedType>,
-        "When publishing by value (i.e. when calling publish(const T& msg)), the published message"
-        "type must not have an overloaded operator new. In this case, please use the"
-        "publish(std::unique_ptr<T> msg) method instead.");
+    if constexpr (detail::has_overloaded_operator_new_v<PublishedType>) {
+      detail::warn_for_overloaded_operator_new();
+    }
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);
     return std::unique_ptr<PublishedType, PublishedTypeDeleter>(ptr, published_type_deleter_);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -52,7 +52,6 @@
 namespace rclcpp
 {
 
-
 template<typename MessageT, typename AllocatorT>
 class LoanedMessage;
 
@@ -87,7 +86,6 @@ public:
 
   /// MessageT::custom_type if MessageT is a TypeAdapter, otherwise just MessageT.
   using PublishedType = typename rclcpp::TypeAdapter<MessageT>::custom_type;
-
   using ROSMessageType = typename rclcpp::TypeAdapter<MessageT>::ros_message_type;
 
   using PublishedTypeAllocatorTraits = allocator::AllocRebind<PublishedType, AllocatorT>;

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -52,6 +52,7 @@
 namespace rclcpp
 {
 
+
 template<typename MessageT, typename AllocatorT>
 class LoanedMessage;
 
@@ -576,7 +577,8 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>::warn();
+    DeprecationEmitterOverloadedOperatorNew<has_overloaded_operator_new_v<
+        PublishedType>>::warn();
 
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -576,7 +576,7 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>::warn();    
+    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>::warn();
 
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -576,7 +576,7 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>{};
+    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>::warn();    
 
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -578,8 +578,10 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<
-        PublishedType>>::warn();
+    static_assert(!detail::has_overloaded_operator_new_v<PublishedType>,
+        "When publishing by value (i.e. when calling publish(const T& msg)), the published "
+    "message type must not have an overloaded operator new. In this case, please use the "
+    "publish(std::unique_ptr<T> msg) method instead.");
 
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -574,6 +574,9 @@ protected:
   std::unique_ptr<PublishedType, PublishedTypeDeleter>
   duplicate_type_adapt_message_as_unique_ptr(const PublishedType & msg)
   {
+    /// Assert that the published type has no overloaded operator new since this leads to new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
+    static_assert(!detail::has_overloaded_operator_new_v<PublishedType>, 
+        "When publishing by value (i.e. when calling publish(const T& msg)), the published message type must not have an overloaded operator new. In this case, please use the publish(std::unique_ptr<T> msg) method instead.");
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);
     return std::unique_ptr<PublishedType, PublishedTypeDeleter>(ptr, published_type_deleter_);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -574,9 +574,12 @@ protected:
   std::unique_ptr<PublishedType, PublishedTypeDeleter>
   duplicate_type_adapt_message_as_unique_ptr(const PublishedType & msg)
   {
-    /// Assert that the published type has no overloaded operator new since this leads to new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    static_assert(!detail::has_overloaded_operator_new_v<PublishedType>, 
-        "When publishing by value (i.e. when calling publish(const T& msg)), the published message type must not have an overloaded operator new. In this case, please use the publish(std::unique_ptr<T> msg) method instead.");
+    /// Assert that the published type has no overloaded operator new since this leads to
+    /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
+    static_assert(!detail::has_overloaded_operator_new_v<PublishedType>,
+        "When publishing by value (i.e. when calling publish(const T& msg)), the published message"
+        "type must not have an overloaded operator new. In this case, please use the"
+        "publish(std::unique_ptr<T> msg) method instead.");
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);
     return std::unique_ptr<PublishedType, PublishedTypeDeleter>(ptr, published_type_deleter_);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -87,6 +87,7 @@ public:
 
   /// MessageT::custom_type if MessageT is a TypeAdapter, otherwise just MessageT.
   using PublishedType = typename rclcpp::TypeAdapter<MessageT>::custom_type;
+
   using ROSMessageType = typename rclcpp::TypeAdapter<MessageT>::ros_message_type;
 
   using PublishedTypeAllocatorTraits = allocator::AllocRebind<PublishedType, AllocatorT>;
@@ -577,7 +578,7 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    DeprecationEmitterOverloadedOperatorNew<has_overloaded_operator_new_v<
+    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<
         PublishedType>>::warn();
 
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);

--- a/rclcpp/include/rclcpp/publisher.hpp
+++ b/rclcpp/include/rclcpp/publisher.hpp
@@ -576,9 +576,8 @@ protected:
   {
     /// Assert that the published type has no overloaded operator new since this leads to
     /// new/delete mismatch (see https://github.com/ros2/rclcpp/issues/2951)
-    if constexpr (detail::has_overloaded_operator_new_v<PublishedType>) {
-      detail::warn_for_overloaded_operator_new();
-    }
+    detail::DeprecationEmitterOverloadedOperatorNew<detail::has_overloaded_operator_new_v<PublishedType>>{};
+
     auto ptr = PublishedTypeAllocatorTraits::allocate(published_type_allocator_, 1);
     PublishedTypeAllocatorTraits::construct(published_type_allocator_, ptr, msg);
     return std::unique_ptr<PublishedType, PublishedTypeDeleter>(ptr, published_type_deleter_);

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -146,14 +146,20 @@ template<typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
 template<bool cond>
-struct DeprecationEmitterOverloadedOperatorNew {};
+struct DeprecationEmitterOverloadedOperatorNew
+{
+};
 
-template<> struct DeprecationEmitterOverloadedOperatorNew<false> {
+template<>
+struct DeprecationEmitterOverloadedOperatorNew<false>
+{
   static void warn() {}
 };
 
-template<> struct 
-DeprecationEmitterOverloadedOperatorNew<true> {
+template<>
+struct
+DeprecationEmitterOverloadedOperatorNew<true>
+{
   [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
         "type must not have an overloaded operator new. In this case, please use the"
         "publish(std::unique_ptr<T> msg) method instead.")]]

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -136,16 +136,18 @@ struct has_overloaded_operator_new<T, std::void_t<
     decltype(T::operator new(std::size_t()))
   >>: std::true_type {};
 
-/// Aligned operator new (C++17)
+template<typename, typename = void>
+struct has_overloaded_aligned_operator_new : std::false_type {};
 template<typename T>
-struct has_overloaded_operator_new<T,
+struct has_overloaded_aligned_operator_new<T,
   std::void_t<decltype( T::operator new(std::size_t(), std::align_val_t()) )>>
   : std::true_type {};
 
 template<typename T>
-inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
+inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value ||
+  has_overloaded_aligned_operator_new<T>::value;
 
-template<bool cond>
+template<bool>
 struct DeprecationEmitterOverloadedOperatorNew
 {
 };
@@ -160,9 +162,9 @@ template<>
 struct
 DeprecationEmitterOverloadedOperatorNew<true>
 {
-  [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
-        "type must not have an overloaded operator new. In this case, please use the"
-        "publish(std::unique_ptr<T> msg) method instead.")]]
+  [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published "
+    "message type must not have an overloaded operator new. In this case, please use the "
+    "publish(std::unique_ptr<T> msg) method instead.")]]
   static void warn() {}
 };
 

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -16,6 +16,7 @@
 #define RCLCPP__TYPE_ADAPTER_HPP_
 
 #include <type_traits>
+#include <new> // For std::align_val_t
 
 namespace rclcpp
 {
@@ -127,6 +128,22 @@ struct assert_type_pair_is_specialized_type_adapter
     type_adapter::is_specialized::value,
     "No type adapter for this custom type/ros message type pair");
 };
+
+template <typename, typename = void>
+struct has_overloaded_operator_new : std::false_type {};
+template <typename T>
+struct has_overloaded_operator_new<T, std::void_t<
+    decltype(T::operator new(std::size_t()))  
+>> : std::true_type {};
+
+/// Aligned operator new (C++17)
+template<class T>
+struct has_overloaded_operator_new<T,
+    std::void_t<decltype( T::operator new(std::size_t(), std::align_val_t()) )>>
+  : std::true_type {};
+
+template <typename T>
+inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
 }  // namespace detail
 

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -147,27 +147,6 @@ template<typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value ||
   has_overloaded_aligned_operator_new<T>::value;
 
-template<bool>
-struct DeprecationEmitterOverloadedOperatorNew
-{
-};
-
-template<>
-struct DeprecationEmitterOverloadedOperatorNew<false>
-{
-  static void warn() {}
-};
-
-template<>
-struct
-DeprecationEmitterOverloadedOperatorNew<true>
-{
-  [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published "
-    "message type must not have an overloaded operator new. In this case, please use the "
-    "publish(std::unique_ptr<T> msg) method instead.")]]
-  static void warn() {}
-};
-
 }  // namespace detail
 
 /// Template metafunction that can make the type being adapted explicit.

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -147,12 +147,18 @@ inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_ne
 
 template<bool cond>
 struct DeprecationEmitterOverloadedOperatorNew {};
-template<> struct DeprecationEmitterOverloadedOperatorNew<false> {};
+
+template<> struct DeprecationEmitterOverloadedOperatorNew<false> {
+  static void warn() {}
+};
+
 template<> struct 
-[[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
+DeprecationEmitterOverloadedOperatorNew<true> {
+  [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
         "type must not have an overloaded operator new. In this case, please use the"
         "publish(std::unique_ptr<T> msg) method instead.")]]
-DeprecationEmitterOverloadedOperatorNew<true> {};
+  static void warn() {}
+};
 
 }  // namespace detail
 

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -145,6 +145,11 @@ struct has_overloaded_operator_new<T,
 template<typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
+[[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
+        "type must not have an overloaded operator new. In this case, please use the"
+        "publish(std::unique_ptr<T> msg) method instead.")]]
+static void warn_for_overloaded_operator_new() {}
+
 }  // namespace detail
 
 /// Template metafunction that can make the type being adapted explicit.

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -137,7 +137,7 @@ struct has_overloaded_operator_new<T, std::void_t<
   >>: std::true_type {};
 
 /// Aligned operator new (C++17)
-template<class T>
+template<typename T>
 struct has_overloaded_operator_new<T,
   std::void_t<decltype( T::operator new(std::size_t(), std::align_val_t()) )>>
   : std::true_type {};
@@ -145,10 +145,14 @@ struct has_overloaded_operator_new<T,
 template<typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
+template<bool cond>
+struct DeprecationEmitterOverloadedOperatorNew {};
+template<> struct DeprecationEmitterOverloadedOperatorNew<false> {};
+template<> struct 
 [[deprecated("When publishing by value (i.e. when calling publish(const T& msg)), the published message"
         "type must not have an overloaded operator new. In this case, please use the"
         "publish(std::unique_ptr<T> msg) method instead.")]]
-static void warn_for_overloaded_operator_new() {}
+DeprecationEmitterOverloadedOperatorNew<true> {};
 
 }  // namespace detail
 

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -16,7 +16,7 @@
 #define RCLCPP__TYPE_ADAPTER_HPP_
 
 #include <type_traits>
-#include <new> // For std::align_val_t
+#include <new>
 
 namespace rclcpp
 {
@@ -129,20 +129,20 @@ struct assert_type_pair_is_specialized_type_adapter
     "No type adapter for this custom type/ros message type pair");
 };
 
-template <typename, typename = void>
+template<typename, typename = void>
 struct has_overloaded_operator_new : std::false_type {};
-template <typename T>
+template<typename T>
 struct has_overloaded_operator_new<T, std::void_t<
-    decltype(T::operator new(std::size_t()))  
->> : std::true_type {};
+    decltype(T::operator new(std::size_t()))
+  >>: std::true_type {};
 
 /// Aligned operator new (C++17)
 template<class T>
 struct has_overloaded_operator_new<T,
-    std::void_t<decltype( T::operator new(std::size_t(), std::align_val_t()) )>>
+  std::void_t<decltype( T::operator new(std::size_t(), std::align_val_t()) )>>
   : std::true_type {};
 
-template <typename T>
+template<typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
 }  // namespace detail

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -143,17 +143,6 @@ struct has_overloaded_operator_new<T,
   : std::true_type {};
 
 template <typename T>
-struct has_overloaded_operator_new<T, std::void_t<
-    decltype(T::operator new[](std::size_t()))  
->> : std::true_type {};
-
-/// Aligned array operator new (C++17)
-template<class T>
-struct has_overloaded_operator_new<T,
-    std::void_t<decltype( T::operator new[](std::size_t(), std::align_val_t()) )>>
-  : std::true_type {};
-
-template <typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
 }  // namespace detail

--- a/rclcpp/include/rclcpp/type_adapter.hpp
+++ b/rclcpp/include/rclcpp/type_adapter.hpp
@@ -143,6 +143,17 @@ struct has_overloaded_operator_new<T,
   : std::true_type {};
 
 template <typename T>
+struct has_overloaded_operator_new<T, std::void_t<
+    decltype(T::operator new[](std::size_t()))  
+>> : std::true_type {};
+
+/// Aligned array operator new (C++17)
+template<class T>
+struct has_overloaded_operator_new<T,
+    std::void_t<decltype( T::operator new[](std::size_t(), std::align_val_t()) )>>
+  : std::true_type {};
+
+template <typename T>
 inline constexpr bool has_overloaded_operator_new_v = has_overloaded_operator_new<T>::value;
 
 }  // namespace detail

